fix(terser): support pure functions for nth_identifier

wmertens opened the pull request:
### Description

Fixes #17409

wmertens committed:
fix(terser): nth_identifier: allow pure functions

yyx990803 commented on packages/vite/src/node/plugins/terser.ts:
```diff
@@ -47,6 +62,16 @@ export function terserPlugin(config: ResolvedConfig): Plugin {
+            | Terser.WeightedIdentifierMangler
+            | undefined = (options.mangle as any)?.nth_identifier
+          if (nth && typeof nth === 'object') {
+            toFunction(nth, 'get')
```
It's kinda surprising to me that the tests passed with this change - it looks like we don't have a test case covering terser minification. /cc @sapphi-red

wmertens replied:
:sweat: sorry I didn't test it, I was hoping the CI would catch any errors. I'll add a test case for both n_th options

yyx990803 commented on packages/vite/src/node/plugins/terser.ts:
```diff
@@ -81,12 +106,25 @@ export function terserPlugin(config: ResolvedConfig): Plugin {
       worker ||= makeWorker()
 
       const terserPath = loadTerserPath(config.root)
+      const nth =
```
Looking at terser option types, it seems `mangle.properties` can also contain `nth` functions: https://github.com/vitejs/vite/blob/main/packages/vite/src/types/terser.d.ts#L118

I was going to propose a more generic function serialize / revive mechanism, but it looks like these are the only two places where functions may appear in all terser options.

@sapphi-red are you aware of other places where we use workers this way and may require handling functions?

wmertens commented:
Ok I added a test and moved the `toFunction` into the worker, but unfortunately I can't make it work.

I don't understand how `parentFunctions` is supposed to work; also, do I understand correctly that when `shouldUseFake` is true, it runs the entire worker on the main thread?

wmertens commented:
Another option would be to augment the terser options with `reproducibleNames` which would then give terser a function that generates a-zA-Z etc, because that's all I'm interested in, reproducible builds.

Would that be accepted?

sapphi-red commented:
> Ok I added a test and moved the `toFunction` into the worker, but unfortunately I can't make it work.

The test is failing because an unrelated test relies on the file structure. It's fine to update the snapshot by `pnpm run test-unit -u`.

> I don't understand how `parentFunctions` is supposed to work

When you pass a function to `parentFunctions`, it will make it possible to call that function inside the worker. It currently requires the function to be async, so it won't work in this case.

> do I understand correctly that when `shouldUseFake` is true, it runs the entire worker on the main thread?

Yes.

> Another option would be to augment the terser options with `reproducibleNames` which would then give terser a function that generates a-zA-Z etc, because that's all I'm interested in, reproducible builds.

If terser isn't deterministic, it's a bug in terser (https://github.com/terser/terser/issues/139#issuecomment-432723143). Would you explain what that option does?

sapphi-red commented:
I'll close this PR as #18987 is merged
